forest/src/lib.rs

```diff
@@ -1,12 +1,605 @@
+    pub fn parent(self, f: &Forest<D, L>) -> Option<Node<D, L>> {
+        f.get(self).parent
+    }
+
+    /// Get the root of the tree containing this node. (This is the same as
+    /// calling `parent()` repeatedly.)
+    pub fn root(self, f: &Forest<D, L>) -> Node<D, L> {
+        let mut root = self;
+        while let Some(parent) = root.parent(f) {
+            root = parent;
+        }
+        root
+    }
+
+    /// Get the data for this node.
+    pub fn data(self, f: &Forest<D, L>) -> &D {
+        &f.get(self).data
+    }
+
+    /// Mutably get the data for this node.
+    pub fn data_mut(self, f: &mut Forest<D, L>) -> &mut D {
+        &mut f.get_mut(self).data
+    }
+
+    /// Return `true` if this is a leaf node (containing `L`), or `false` if it's
+    /// a branch node (containing children).
+    pub fn is_leaf(self, f: &Forest<D, L>) -> bool {
+        matches!(f.get(self).leaf_or_children, Leaf(_))
+    }
+
+    /// Get the leaf data at this leaf node.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this node is not a leaf.
+    pub fn leaf(self, f: &Forest<D, L>) -> &L {
+        f.leaf(self)
+    }
+
+    /// Mutably get the data at this leaf node.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this node is not a leaf.
+    pub fn leaf_mut(self, f: &mut Forest<D, L>) -> &mut L {
+        f.leaf_mut(self)
+    }
+
+    /// Get this node's children.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this is not a branch node.
+    pub fn children(self, f: &Forest<D, L>) -> &[Node<D, L>] {
+        f.children(self)
+    }
+
+    /// This node's siblings, in order, including itself.
+    /// If this is a root node, its siblings are the roots.
+    pub fn siblings(self, f: &Forest<D, L>) -> &[Node<D, L>] {
+        f.siblings(self)
+    }
+
+    /// Determine this node’s index among its siblings.
+    pub fn sibling_index(self, f: &Forest<D, L>) -> usize {
+        f.sibling_index(self)
+    }
+
+    /// Detach this node from its parent. Afterwards, it will be a root node,
+    /// and its parent will have one fewer child. If this node was already a
+    /// root node, this is a no-op.
+    pub fn detach(self, f: &mut Forest<D, L>) {
+        if let Some(parent) = f.get(self).parent {
+            let i = f.sibling_index(self);
+            f.children_mut(parent).remove(i);
+            f.get_mut(self).parent = None;
+            f.roots.push(self);
+        }
+    }
+
+    /// Insert `node` as this node's i'th child, after removing it from any
+    /// previous parent it may have had.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `child_index` is out of bounds, or if this node isn't a branch,
+    /// or if you're trying to make a cycle like a douchebag.
+    pub fn insert_child(self, f: &mut Forest<D, L>, child_index: usize, node: Node<D, L>) {
+        if self.root(f) == node.root(f) {
+            panic!("Forest - attempt to create cycle using `insert_child` thwarted");
```

Comment:
Require that the node to be inserted is a root, but not the root of its new parent. Document that you need to `detach` the node before calling `insert_child`, if it's not already a root. This avoids confusion about `child_index` when re-inserting a child at the same parent but with a different index than before.